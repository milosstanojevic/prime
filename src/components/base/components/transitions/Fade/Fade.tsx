import React, { FC, ReactElement, cloneElement } from 'react';
import { Transition } from 'react-transition-group';
import { TransitionProps } from '../types';

interface IFadeProps extends Omit<TransitionProps, 'timeout'> {
    children: ReactElement;
    timeout?: number;
}

const defaultStyle = (duration: number): Partial<{ transition: string; opacity: number }> => ({
    transition: `opacity ${duration}ms ease-in-out`,
    opacity: 0
});

const transitionStyles = {
    entering: { opacity: 1 },
    entered: { opacity: 1 },
    exiting: { opacity: 0 },
    exited: { opacity: 0 }
};

export const Fade: FC<IFadeProps> = ({
    in: inProp = false,
    onEnter,
    onExited,
    children,
    style,
    timeout = 500
}) => (
    <Transition appear in={inProp} timeout={timeout} onEnter={onEnter} onExited={onExited}>
        {(state): ReactElement => {
            return cloneElement(children, {
                style: {
                    visibility: state === 'exited' && !inProp ? 'hidden' : undefined,
                    ...defaultStyle(timeout),
                    // @ts-ignore
                    ...transitionStyles[state],
                    ...style,
                    ...children.props.style
                }
            });
        }}
    </Transition>
);

Fade.displayName = 'Fade';
