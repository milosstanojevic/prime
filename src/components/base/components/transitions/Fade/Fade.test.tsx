import React from "react"
import { render } from "enzyme"
import { Fade } from "./Fade"

describe("<Fade />", () => {
  it("should render", () => {
    expect(render(
      <Fade>
        <span>Test</span>
      </Fade>
    )).toMatchSnapshot();
  });

  it("should render as entered", () => {
    expect(render(
      <Fade in timeout={300}>
        <span>Test</span>
      </Fade>
    )).toMatchSnapshot();
  });
});
