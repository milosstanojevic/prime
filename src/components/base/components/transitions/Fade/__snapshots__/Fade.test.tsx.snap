// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fade /> should render 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "style": "visibility:hidden;transition:opacity 500ms ease-in-out;opacity:0",
    },
    "children": Array [
      Node {
        "data": "Test",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`<Fade /> should render as entered 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "style": "transition:opacity 300ms ease-in-out;opacity:0",
    },
    "children": Array [
      Node {
        "data": "Test",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
